### DISCUSSION TITLE
ZFS reservation: installation guide vs command

### DISCUSSION THREAD

**Discussion by FStefanni** [openzfs/zfs#15034] [Q&A] [answered]
Hi,

I know that due to performance degradation with full disk, it is better to reserve some disk quota.

In the official installation guides, the tutorials keep 1GiB of free disk at the end of partitions (I mean: 1non-partitioned GB), and a comment says: "At least 1GB".

So my question is: is this just a way to do the reservation for the known performance issues?

If so, is it the same result of doing a reservation with a command? E.g.:
```
zfs create -o refreservation=1G -o mountpoint=none zroot/reserved
```

Thank you.
Regards

**Comment by rincebrain** [comment]
You're conflating a few things here.

ZFS, like all filesystems, when very very full, can sometimes run markedly slower, especially with high churn or space fragmentation, in part because you stop being able to reliably find chunks large enough for whatever you're trying to write, and have to start hacking things into smaller chunks, and that's inefficient in a few ways.

But that's based on free space in the pool itself, not space ZFS doesn't think it controls on the disk. There are a lot of tutorials for ZFS, so I can't guess which ones you specifically looked at or why they were suggesting that.

Setting a reservation or refreservation on a dataset that's doing nothing else would indeed keep a minimum amount free, but ZFS already reserves a chunk of free space to ensure it can always have enough space to free deleted things (among other things), since "never overwriting things in place" means you always need free space in order to do things, and the "slop space" reservation is how OpenZFS avoids being able to fill a filesystem so much you can't delete things due to lack of space.

I believe the slop space is capped in recent releases to no more than 128 GiB, as an upper limit, when previously it was just 3.125% of pool space, with no limit. So that should still be a dominating amount of free space on many pools compared to reserving a GB.

**Comment by IvanVolosyuk** [comment]
Yeah, as Rich mentioned, refreservation is too keep some space unallocated
within the pool. Some people try to reserve more than what ZFS does by
default to avoid GANG block creation, which historically caused slow down
in the pools. That's pretty much is trading space for performance.

One GB outside of pool can be another workaround. It is not possible to
shrink a vdev, but disks from different manufacturers can vary slightly in
size. ZFS already reserves small amount of space for the whole disk vdevs,
so that you can replace it with slightly smaller disk from different
manufacturer. If you partition your disk manually, you can create small
partition anywhere in the disk for the same purpose and blocks at the end
are the slowest for HDDs.

**Comment by FStefanni** [comment]
Hi,

thank you for the clarifications.

FYI, I am looking at these tutorials:
- Method 1: Official guide of ZFS for NixOS: https://openzfs.github.io/openzfs-docs/Getting%20Started/NixOS/Root%20on%20ZFS.html
    * Here, 1GB is left at the end of disk, with the comment of "minimum 1GB" 
- Method 2: NixOS page on tips for ZFS: https://nixos.wiki/wiki/ZFS#Reservations
    * Here the reservation is done by creating an unused dataset

So, if I understand correctly your feedbacks:
* Method 2 works fine, since it actually forces some free space
    * It works for ZFS, since it belongs to an actual pool, whose datasets are all used (except its "reservation" dataset)
    * It is useful for performance, but not mandatory for ZFS
    * It could help to avoid fragmentation/GANG blocks
* Method 1 is not related to reservations
    * It is a trick to ensure disk compatibility between different manufactures
    * E.g. in case in a raid setup, one of them goes broken

Did I understood correctly?

Thank you,
regards

**Reply by rincebrain** [reply, marked as answer]
No.

I was saying that a single GiB is probably useless to reserve because your pool has likely already reserved much more than a GiB for its own reasons, so one more isn't going to make that much difference.

For example, on a 5TiB pool, 3.125% would be 156 GiB, so you'd either be reserving 128 GiB already (if you have a version new enough to limit the slop space that way) or the whole 156 GiB (if not), either of which is much more than 1 GiB.

**Comment by FStefanni** [comment]
Ok now it is clear.
Thank you.
Regards
